packages/server/lib/browsers/cri-client.ts

```diff
@@ -128,17 +129,73 @@ const maybeDebugCdpMessages = (cri) => {
  */
 export { chromeRemoteInterface }
 
-export const create = Bluebird.method((debuggerUrl: websocketUrl): Bluebird<CRIWrapper> => {
-  return chromeRemoteInterface({
-    target: debuggerUrl,
-    local: true,
-  })
-  .then((cri) => {
-    maybeDebugCdpMessages(cri)
+type DeferredPromise = { resolve: Function, reject: Function }
+
+export const create = Bluebird.method((target: websocketUrl, onAsynchronousError: Function): Bluebird<CRIWrapper> => {
+  const subscriptions: {eventName: CRI.EventName, cb: Function}[] = []
+  let enqueuedCommands: {command: CRI.Command, params: any, p: DeferredPromise }[] = []
+
+  let closed = false // has the user called .close on this?
+  let connected = false // is this currently connected to CDP?
+
+  let cri
+  let client: CRIWrapper
+
+  const reconnect = () => {
+    debug('disconnected, attempting to reconnect... %o', { closed })
+
+    connected = false
+
+    if (closed) {
+      return
+    }
+
+    return connect()
+    .then(() => {
+      debug('restoring subscriptions + running queued commands... %o', { subscriptions, enqueuedCommands })
+      subscriptions.forEach((sub) => {
+        cri.on(sub.eventName, sub.cb)
+      })
+
+      enqueuedCommands.forEach((cmd) => {
+        cri.send(cmd.command, cmd.params)
+        .then(cmd.p.resolve, cmd.p.reject)
+      })
```

Comment:
i don't believe the logic here is correctly connected because the `forEach` is synchronous and is not properly returning the mapped promises.

I believe you want to return `Promise.mapSeries` here. I could be wrong though.

Reply:
that's what we want here - the intention of this code is to fire off all of the enqueued commands, in the order they were sent, and then resolving/rejecting the originally returned (deferred) promise.

the `.catch` on line 167 is mostly to ensure that errors from `connect()` are thrown somewhere instead of being unhandled, but it could potentially catch any synchronous errors in cri.on and cri.send as well

Reply:
I suppose that the outer handlers calling into `.send()` are still returned the enqueued promise - which allows them to attach any `.catch()` handlers.

My concern was of uncaught promises - but I believe the chain is correctly propagated. 

I'll approve this - although I should look at the tests in more detail to verify that they're specifically covering the scenario I'm imagining.